python_modules/libraries/dagster-components/dagster_components/core/component.py

```diff
@@ -206,35 +212,69 @@ def get_registered_components_in_module(module: ModuleType) -> Iterable[Type[Com
 T = TypeVar("T")
 
 
+@record
+class ComponentInstanceKey:
```

Comment:
I buy that this needs to exist, but I think we should incorporate a few lessons learned from mild annoyances with the `AssetKey` class:

1. we should be quite strict about the components of this key (each part in the path should be a valid python identifier) -- it's hard to walk this back later on
2. covered by the above, but definitely want to make sure a part of the key can't contain a dot so that there's no confusion when converting back and forth

Reply:
Great feedback

Reply:
In terms of name: I'm tempted to just name it `ComponentKey` for simplicity, which then brings up the real core of the naming discussion, which is how we expect people to verbally differentiate between a "Component Type" and a "Component Instance".

I posit that colloquially, people are going to want to just say "component" for one of the two use cases, and if so, we should make sure that use case is for "component instance", and orient our terminology around that.

My feeling is that people are going to want to talk about component instances way more often than component types, and so we should make component the "correct" word for that use case.

Proposal:

When referring to things in text etc,
- "Component" = "Component Instance"
- "Component Type" = "Component Type" (a different word could be used here in theory)

Ramifications:
- `@component` -> `@component_type`
- `Component` -> `ComponentType`
- `ComponentInstanceKey` -> `ComponentKey`

Sample phrases I think people will want to say:
- "Where does that component live in the UI?"
- "I just added another component to the project"
  - note: ambiguous in current day, could mean either "component type" or "component instance", but people are going to add component instances way more often, hence why I think we should optimize for it
- "I just added another component type to the project"

Reply:
Agree with everything except for the rename of `class Component` to `class ComponentType`. Class implies that it is a type already.